onnxruntime/core/providers/qnn/builder/qnn_backend_manager.cc

```diff
@@ -128,6 +130,45 @@ Status QnnBackendManager::LoadQnnSystemLib() {
   return Status::OK();
 }
 
+void QnnLogging(const char* format,
+                QnnLog_Level_t level,
+                uint64_t timestamp,
+                va_list argument_parameter) {
+  ORT_UNUSED_PARAMETER(level);
+  ORT_UNUSED_PARAMETER(timestamp);
+
+  std::lock_guard<std::mutex> lock(qnn_log_mutex_);
+
+  // Always output Qnn log as Ort verbose log
+  size_t ini_size = 20;
+  std::vector<char> buffer(ini_size);
+  auto length = vsnprintf(buffer.data(), ini_size, format, argument_parameter);
+  buffer.resize(length);
+  vsnprintf(buffer.data(), length, format, argument_parameter);
+  LOGS_DEFAULT(VERBOSE) << buffer.data();
+}
+
+void QnnBackendManager::InitializeQnnLog() {
+  const std::map<logging::Severity, QnnLog_Level_t> ort_log_level_to_qnn_log_level = {
+      {logging::Severity::kVERBOSE, QNN_LOG_LEVEL_DEBUG},
+      {logging::Severity::kINFO, QNN_LOG_LEVEL_INFO},
+      {logging::Severity::kWARNING, QNN_LOG_LEVEL_WARN},
+      {logging::Severity::kERROR, QNN_LOG_LEVEL_ERROR},
+      {logging::Severity::kFATAL, QNN_LOG_LEVEL_ERROR}};
+
+  // Set Qnn log level align with Ort log level
+  QnnLog_Level_t qnn_log_level = QNN_LOG_LEVEL_WARN;
+  auto ort_log_level = logger_->GetSeverity();
+  auto pos = ort_log_level_to_qnn_log_level.find(ort_log_level);
+  if (pos != ort_log_level_to_qnn_log_level.end()) {
+    qnn_log_level = pos->second;
+  }
```

Comment:
Please feel free to ignore this suggestion (this function is only called once).

I would prefer to avoid the use of `std::map` for this simple conversion between enums. `std::map` introduces a heap allocation and seems a bit heavy weight for this task in my opinion. A switch statement avoids heap allocation and is likely compiled down to simple jump table.

```C++
    QnnLog_Level_t qnn_log_level = QNN_LOG_LEVEL_WARN;
    auto ort_log_level = logger_->GetSeverity();

    switch (ort_log_level) {
      case logging::Severity::kVERBOSE:
        qnn_log_level = QNN_LOG_LEVEL_DEBUG;
        break;
      case logging::Severity::kINFO:
        qnn_log_level = QNN_LOG_LEVEL_INFO;
        break;
      case logging::Severity::kWARNING:
        qnn_log_level = QNN_LOG_LEVEL_WARN;
        break;
      case logging::Severity::kERROR:
        qnn_log_level = QNN_LOG_LEVEL_ERROR;
        break;
      default:
        qnn_log_level = QNN_LOG_LEVEL_WARN;
        break;
    }
}
```

Alternatively, we could just use a fixed-sized array, but that requires a few static_casts between the enum and int.
```C++
QnnLog_Level_t ort_log_level_to_qnn_log_level[MAX_ORT_SEVERITY] = { QNN_LOG_LEVEL_DEBUG, ...};
```

I'm a bit keen about reducing heap allocations, but in this case feel free to leave it as is.

Reply:
updated. thanks!

---
In reply to: [1199322703](https://github.com/microsoft/onnxruntime/pull/16019#discussion_r1199322703) [](ancestors = 1199322703)